File: core/causality/cross_runtime_alignment_engine.py
from __future__ import annotations

from typing import Any, Dict, List


RUNTIME_ORDER = (
    "browser",
    "application",
    "electron",
    "desktop",
    "terminal",
    "vm",
    "remote",
    "distributed",
)


def align_cross_runtime_events(
    events: List[Dict[str, Any]],
) -> Dict[str, Any]:
    buckets: Dict[str, List[Dict[str, Any]]] = {
        runtime: [] for runtime in RUNTIME_ORDER
    }

    for event in events[:20000]:
        runtime = str(event.get("runtime", "browser")).lower()
        if runtime not in buckets:
            buckets[runtime] = []
        buckets[runtime].append(event)

    aligned: List[Dict[str, Any]] = []
    step = 0
    for runtime in RUNTIME_ORDER:
        for event in sorted(
            buckets.get(runtime, []),
            key=lambda item: int(item.get("step", 0)),
        ):
            aligned.append({
                **event,
                "aligned_step": step,
                "aligned_runtime": runtime,
            })
            step += 1

    correlations = []
    for index in range(1, len(aligned)):
        prev = aligned[index - 1]
        curr = aligned[index]
        correlations.append({
            "from_runtime": prev.get("runtime", ""),
            "to_runtime": curr.get("runtime", ""),
            "from_event": prev.get("id", ""),
            "to_event": curr.get("id", ""),
            "correlation_id": f"corr:{index}",
        })

    return {
        "aligned_events": aligned,
        "correlations": correlations,
        "runtime_count": len({e.get("runtime") for e in aligned}),
        "bounded": True,
    }

File: src/causality/crossRuntimeAlignmentEngine.ts
/**
 * Converted from Python: core/causality/cross_runtime_alignment_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";

export let RUNTIME_ORDER: any = ["browser", "application", "electron", "desktop", "terminal", "vm", "remote", "distributed"];
export function alignCrossRuntimeEvents(events: any): any {
  var buckets: any = Object.fromEntries(py.iter(RUNTIME_ORDER).map((runtime: any) => ([runtime, []] as [any, any])));
  var event: any;
  for (event of py.iter(py.slice(events, null, 20000))) {
    var runtime: any = String(py.toStr(py.get(event, "runtime", "browser"))).toLowerCase();
    if (!py.contains(buckets, runtime)) {
      py.setItem(buckets, runtime, []);
    }
    py.listAppend(py.at(buckets, runtime), event);
  }
  var aligned: any[] = [];
  var step: any = 0;
  for (runtime of py.iter(RUNTIME_ORDER)) {
    for (event of py.iter(py.sorted(py.get(buckets, runtime, []), {key: ((item: any) => py.toInt(py.get(item, "step", 0))) as (item: any) => any}))) {
      py.listAppend(aligned, {...(event), "aligned_step": step, "aligned_runtime": runtime});
      step = py.add(step, 1);
    }
  }
  var correlations: any[] = [];
  var index: any;
  for (index = 1; index < py.len(aligned); index++) {
    var prev: any = py.at(aligned, py.sub(index, 1));
    var curr: any = py.at(aligned, index);
    py.listAppend(correlations, {"from_runtime": py.get(prev, "runtime", ""), "to_runtime": py.get(curr, "runtime", ""), "from_event": py.get(prev, "id", ""), "to_event": py.get(curr, "id", ""), "correlation_id": `corr:${py.toStr(index)}`});
  }
  return {"aligned_events": aligned, "correlations": correlations, "runtime_count": py.len(py.toSet(py.iter(aligned).map((e: any) => py.get(e, "runtime")))), "bounded": true};
}
